google-cloud-storage/src/main/java/com/google/cloud/storage/StorageImpl.java

```diff
@@ -733,6 +736,124 @@ public URL signUrl(BlobInfo blobInfo, long duration, TimeUnit unit, SignUrlOptio
+    googDateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+    yearMonthDayFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+    expirationFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+
+    long timestamp = getOptions().getClock().millisTime();
+    String date = googDateFormat.format(timestamp);
+    String signingCredential =
+        credentials.getAccount()
+            + "/"
+            + yearMonthDayFormat.format(timestamp)
+            + "/auto/storage/goog4_request";
+
+    Map<String, String> policyFields = new HashMap<>();
+
+    V4PostConditions.Builder conditionsBuilder = conditions.toBuilder();
+
+    for (Map.Entry<String, String> entry : fields.getFieldsMap().entrySet()) {
+      // Every field needs a corresponding policy condition, so add them if they're missing
+      conditionsBuilder.addCustomCondition(
+          V4ConditionType.MATCHES, entry.getKey(), entry.getValue());
+
+      policyFields.put(entry.getKey(), entry.getValue());
+    }
+
+    V4PostConditions v4Conditions =
+        conditionsBuilder
+            .addKeyCondition(V4ConditionType.MATCHES, escapedBlobName)
+            .addCustomCondition(V4ConditionType.MATCHES, "x-goog-date", date)
+            .addCustomCondition(V4ConditionType.MATCHES, "x-goog-credential", signingCredential)
+            .addCustomCondition(V4ConditionType.MATCHES, "x-goog-algorithm", "GOOG4-RSA-SHA256")
+            .build();
+    V4PostPolicyDocument document =
+        V4PostPolicyDocument.of(
+            expirationFormat.format(
+                timestamp + TimeUnit.MILLISECONDS.convert(duration, TimeUnit.SECONDS)),
+            v4Conditions);
+    String policy = BaseEncoding.base64().encode(document.toJsonObject().getBytes());
+    String signature =
+        BaseEncoding.base16().encode(credentials.sign(policy.getBytes())).toLowerCase();
+
+    for (V4Condition condition : v4Conditions.getConditions()) {
+      if (condition.type == V4ConditionType.MATCHES) {
+        policyFields.put(condition.element, condition.value);
+      }
+    }
+    policyFields.put("key", escapedBlobName);
+    policyFields.put("x-goog-credential", signingCredential);
+    policyFields.put("x-goog-algorithm", "GOOG4-RSA-SHA256");
+    policyFields.put("x-goog-date", date);
+    policyFields.put("x-goog-signature", signature);
+    policyFields.put("policy", policy);
+
+    return V4PostPolicy.of(url, policyFields);
+  }
+
+  public V4PostPolicy generateV4PresignedPostPolicy(
+      BlobInfo blobInfo, V4PostFields fields, long duration, V4PostPolicyOption... options) {
+    return generateV4PresignedPostPolicy(
+        blobInfo, fields, V4PostConditions.newBuilder().build(), duration, options);
+  }
+
+  public V4PostPolicy generateV4PresignedPostPolicy(
+      BlobInfo blobInfo, long duration, V4PostPolicyOption... options) {
+    return generateV4PresignedPostPolicy(
+        blobInfo, V4PostFields.newBuilder().build(), duration, options);
+  }
+
```

Comment:
Put V4 at the end of the name please.

Reply:
Please do this for other helpers as well that use V4 for POST Policy. From meeting.